### PR TITLE
Use vtkTransformFilter in DataSet.transform()

### PR DESCRIPTION
**Pull request by whophil**
### Overview

 

In trying to actually use some of the features added in #1127, I realized that the transformations in PyVista are for geometry only, and don't handle transformations of any vector/tensor arrays. This is true of `DataSet.rotate_x()`, `DataSet.rotate_y()`, `DataSet.rotate_z()`, and `DataSet.transform()`, all of which just transform points.

While we could transform vector fields manually, [`vtkTransformFilter`](https://vtk.org/doc/nightly/html/classvtkTransformFilter.html) also does. Is there a good reason not to use it?

This PR shows that might look like for `DataSet.transform()`. Some logic could be cleaned up a bit by creating a helper class to manage the various 4x4 `ndarray`/`vtkMatrix`/`vtkTransform` types that might be passed in. There's also work to do in implementing this across the other transform methods, but I wanted to get some of your thoughts before proceeding (@akaszynski ?).

### CODE REVIEW

**Comment by akaszynski** [comment]
Picking the best of both worlds works in this case, and thanks for thinking of this.

On a side note:
I ran into an issue the other day with `axis_rotation` because it no longer supports individual angles for each point, but one global rotation.  I don't think this is something we have to support, but we can consider adding back in the old rotation method for point arrays.

One of the reason why VTK is so powerful is because they generally account for a wide variety of edge cases, and this is why it's hard to implement our own and make it robust.

**Comment by whophil** [comment]
This is ready for review. The static code check is failing in a way I don't understand; there are plenty of other methods inside filters.py which use `dataset.overwrite()`. 

Also, the doc build is failing with
```
AttributeError: 'vtkFiltersGeneralPython.vtkTransformFilter' object has no attribute 'SetTransformAllInputVectors'
```
when this is exercised and passing in the other test jobs. Is there something different about the environment for BuildDocumentation?

**Comment by akaszynski** [comment]
@whophil, I'll look at the failures.

**Comment by akaszynski** [comment]
Issue with `mypy` fixed by 834c434b7422d52990e752e83b40155afcef9897.  Specifically, you have to define the type upfront.  Since we can input a variety of inputs, this type def check is actually helpful.

**Comment by akaszynski** [comment]
> Also, the doc build is failing with
> 
> ```
> AttributeError: 'vtkFiltersGeneralPython.vtkTransformFilter' object has no attribute 'SetTransformAllInputVectors'
> ```
> 
> I'm confused because this method is exercised in the other test jobs which are passing. Is there something different about the environment for BuildDocumentation?

Yes, we're using `vtk==8.1.2`, which I think is causing the issue.  In fact, it fails immediately when collect pytest with:
```
_________________ ERROR collecting tests/test_itk_plotting.py _________________
tests/test_itk_plotting.py:20: in <module>
    SPHERE = pyvista.Sphere()
pyvista/utilities/geometric_objects.py:289: in Sphere
    surf.rotate_y(-90)
pyvista/core/dataset.py:482: in rotate_y
    self.transform(t, transform_all_input_vectors=transform_all_input_vectors, inplace=True)
pyvista/core/filters.py:2690: in transform
    f.SetTransformAllInputVectors(transform_all_input_vectors)
E   AttributeError: 'vtkFiltersGeneralPython.vtkTransformFilter' object has no attribute 'SetTransformAllInputVectors'
```

**Comment by whophil** [comment]
> @whophil, I've got a venv setup for python3.7 locally. Do you want me to investigate this, or do you think you got it?

Ah ok. I will push a fix to make this compatible with 8.1.2. Thanks for the help!

**Comment by akaszynski** [comment]
> Ah ok. I will push a fix to make this compatible with 8.1.2. Thanks for the help!

No problem.  It's a bit of a pain supporting 8.1.2 as well as 9+, but I think it's worth it considering that we're trying to make vtk easier to use, and 9.0 has some issues (e.g. plots not closing on Linux).

**Comment by whophil** [comment]
@akaszynski this is ready to go now. I added a `VTKVersionError` to `pyvista.core.errors`, feel free to rework or expound upon if you see fit.

The transform tests have been modified to run on VTK 8.1.2 even if that isn't a part of CI. With VTK 8.1.2, the tests are passing locally.

File: pyvista/core/filters.py [hunk 1]
```diff
@@ -2638,7 +2640,82 @@ def shrink(dataset, shrink_factor=1.0, progress_bar=False):
...
+
+        Parameters
+        ----------
+        trans : vtk.vtkMatrix4x4, vtk.vtkTransform, or np.ndarray
+            Accepts a vtk transformation object or a 4x4 transformation matrix.
+
+        transform_all_input_vectors: bool, optional
+            When ``True``, all input vectors are transformed. Otherwise, only the
+            points, normals and active vectors are transformed.
+        """
```
Comment (akaszynski) [inline pyvista/core/filters.py:2656]: Let's add an `Example` here, and I think we'll be good to go.  It really helps with the sphinx autodocs.

**Review by akaszynski — APPROVED** [review]
Perfect.  Thanks!

**Pull request merged**